pkg/csource/options.go

```diff
@@ -8,10 +8,91 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"sort"
+	"strings"
 
 	"github.com/google/syzkaller/pkg/mgrconfig"
 )
 
+type Feature struct {
+	Name        string
+	Description string
+	Enabled     bool
+}
+
+type Features map[string]Feature
+
+func defaultFeatures(value bool) Features {
+	return map[string]Feature{
+		"tun":         {"tun", "setup and use /dev/tun for packet injection", value},
+		"net_dev":     {"net_dev", "setup more network devices for testing", value},
+		"net_reset":   {"net_reset", "reset network namespace between programs", value},
+		"cgroups":     {"cgroups", "setup cgroups for testing", value},
+		"binfmt_misc": {"binfmt_mist", "setup binfmt_misc for testing", value},
+	}
+}
+
+func enableAllFeatures(features Features) Features {
+	for name, feature := range features {
+		feature.Enabled = true
+		features[name] = feature
+	}
+	return features
+}
+
+func disableAllFeatures(features Features) Features {
+	for name, feature := range features {
+		feature.Enabled = false
+		features[name] = feature
+	}
+	return features
+}
+
+func ParseFeaturesFlags(enable string, disable string, defaultValue bool) (Features, error) {
+	if enable == "none" && disable == "none" {
+		return defaultFeatures(defaultValue), nil
+	}
+	if enable != "none" && disable != "none" {
+		return nil, fmt.Errorf("can't use -enable and -disable flags at the same time")
+	}
+	if enable == "all" || disable == "" {
+		return enableAllFeatures(defaultFeatures(defaultValue)), nil
+	}
+	if disable == "all" || enable == "" {
+		return disableAllFeatures(defaultFeatures(defaultValue)), nil
+	}
+	var items []string
+	var features Features
+	if enable != "none" {
+		items = strings.Split(enable, ",")
+		features = disableAllFeatures(defaultFeatures(defaultValue))
+	} else {
+		items = strings.Split(disable, ",")
+		features = enableAllFeatures(defaultFeatures(defaultValue))
+	}
+	for _, item := range items {
+		if _, ok := features[item]; !ok {
+			var names []string
+			for name := range features {
+				names = append(names, name)
+			}
+			sort.Strings(names)
+			var featuresList string
+			for i, name := range names {
+				featuresList += fmt.Sprintf("  %s - %s", name, features[name].Description)
+				if i != len(names)-1 {
+					featuresList += "\n"
+				}
+			}
+			return nil, fmt.Errorf("unknown feature specified: %s\n%s", item, featuresList)
+		}
+		feature := features[item]
+		feature.Enabled = (enable != "none")
+		features[item] = feature
+	}
+	return features, nil
+}
+
 // Options control various aspects of source generation.
 // Dashboard also provides serialized Options along with syzkaller reproducers.
 type Options struct {
```

Comment:
Please keep this close to the top of the file. This file is about csource Options, when one opens it they want to see the options.

Reply:
Done.